sematic/db/models/external_resource_record.py

```diff
@@ -0,0 +1,147 @@
+from sematic.db.models.json_encodable_mixin import JSONEncodableMixin
+from sematic.external_resource import ExternalResource
+from sematic.types.serialization import (
+    type_from_json_encodable,
+    type_to_json_encodable,
+    value_from_json_encodable,
+    value_to_json_encodable,
+)
+
+TypeSerialization = Dict[str, Any]
+
+
+class ExternalResourceRecord(Base, JSONEncodableMixin):
+    """A DB record for an ExternalResource and its history
+
+    Attributes
+    ----------
+    id:
+        The unique id of the external resource
+    resource_state:
+        The current state of the resource
+    status_message:
+        The most recent status message for the resource
+    last_updated_epoch_seconds:
+        The time that the resource was last updated, expressed as epoch seconds
+    type_serialization:
+        The Sematic serialization of the type. This may be different from the
+        serialization for ExternalResource itself, since the resource will be a
+        subclass of ExternalResource.
+    value_serialization:
+        The Sematic serialization of the object. This will contain extra fields
+        unique to the subclasses of ExternalResource. It also does contain
+        json variants of some of the status fields above, but the duplicated values
+        should match.
+    history_serializations:
+        The Sematic serialization of the objects. Any time the ExternalResource object
+        changes in such a way as to make the instances compare as not equal, a new
+        entry will be added to this list. Element 0 is the most recent, element N
+        the oldest.
+    """
+
+    # Q: Why duplicate data that's already in the json of value_serialization as columns?
+    # A: For two reasons:
+    #    1. It gives us freedom to refactor the dataclass for ExternalResource later
+    #    to move fields around, while allowing the database columns to stay stable.
+    #    2. It allows for more efficient queries on the explicit columns rather than
+    #    requiring json traversal.
+
+    __tablename__ = "external_resources"
+
+    id: str = Column(types.String(), primary_key=True)
+    resource_state: str = Column(  # type: ignore
+        types.String(),
+        nullable=False,
+    )
+    status_message: str = Column(types.String(), nullable=False)
+    last_updated_epoch_seconds: int = Column(types.BIGINT(), nullable=False)
+    type_serialization: TypeSerialization = Column(types.JSON(), nullable=False)
+    value_serialization: Dict[str, Any] = Column(types.JSON(), nullable=False)
+    history_serializations: Tuple[Dict[str, Any], ...] = Column(  # type: ignore
+        types.JSON(), nullable=False
+    )
+
+    @classmethod
+    def from_resource(cls, resource: ExternalResource) -> "ExternalResourceRecord":
+        if not isinstance(resource, ExternalResource):
+            raise ValueError(
+                f"resource must be an instance of a subclass of "
+                f"ExternalResource. Was: {resource}"
```

Comment:
The repr for the effective value might hide important information.
```suggestion
                f"ExternalResource. Was: {resource} of type {type(resource)}."
```